### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by StuartCowley**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,18 @@
-# Getting Started with Create React App
+NASA Image Search
```
Comment (StuartCowley) [inline README.md:1]: The heading of you README should have a single hash at the start:
```
# NASA Image Search
```

File: README.md [hunk 2]
```diff
@@ -1,4 +1,18 @@
...
+=================
+
+This is my second project produced with React. It is an image search application and uses a NASA image API to display images after a successful search. The images are displayed using [https://github.com/xiaolin/react-image-gallery](https://github.com/xiaolin/react-image-gallery), a React carousel image gallery component with thumbnail support. Documentation for further experimentation can be found at [https://github.com/xiaolin/react-image-gallery/blob/master/README.md](https://github.com/xiaolin/react-image-gallery/blob/master/README.md)
+
+Further features to include:
+
+- full screen image viewing
+- gallery customisation such as index on/off, bullet navigation, image size choices, addtional information pulled from the NASA api and displayed for the user.
```
Comment (StuartCowley) [inline README.md:9]: API should be capitalised here:
```
- gallery customisation such as index on/off, bullet navigation, image size choices, addtional information pulled from the NASA API and displayed for the user.
```

File: README.md [hunk 3]
```diff
@@ -1,4 +1,18 @@
-# Getting Started with Create React App
+NASA Image Search
+=================
+
+This is my second project produced with React. It is an image search application and uses a NASA image API to display images after a successful search. The images are displayed using [https://github.com/xiaolin/react-image-gallery](https://github.com/xiaolin/react-image-gallery), a React carousel image gallery component with thumbnail support. Documentation for further experimentation can be found at [https://github.com/xiaolin/react-image-gallery/blob/master/README.md](https://github.com/xiaolin/react-image-gallery/blob/master/README.md)
```
Comment (StuartCowley) [inline README.md:4]: This is a great start to your README, it would be great if you could add:

- a set of installation instructions,
- tech used,
- how to test it,
- a replacement for the boilerplate instructions (such as the section about `npm start` and tweak it to be in your own words),
- your name!

Good to see the stretch goals here as well as the accreditation for the external library, Id add another here about making the app responsive for mobile devices too

File: _code/code.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+          <div className="image-container">
```
Comment (StuartCowley) [inline _code/code.js:1]: Does this file need to be here? It seems to be left over from something else, if it serves no purpose it should be removed

File: src/styles/searchResults.css [hunk 5]
```diff
@@ -0,0 +1,15 @@
+#results-heading {
```
Comment (StuartCowley) [inline src/styles/searchResults.css:1]: Targeting IDs for CSS classes isnt really recommended, they are usually for finding elements with JS and using classnames as the primary method of lookup for CSS. There are a few edge cases but since this isnt one of them, you should replace this with a `.results-heading` class

File: src/__tests__/App.test.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+import { render, screen } from "@testing-library/react";
```
Comment (StuartCowley) [inline src/__tests__/App.test.js:1]: You dont need the `screen` method here, you can delete it
```
import { render } from "@testing-library/react";
```

File: src/__tests__/App.test.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
...
+import App from "../components/App";
+
+describe('App', () => {
+  const { asFragment } = render(<App />)
+
+  test("app renders correctly", () => {
+    expect(asFragment).toMatchSnapshot()
```
Comment (StuartCowley) [inline src/__tests__/App.test.js:8]: If you look at the snapshot generated by this test, you should expect to see a layout of a component complete with divs, text rendered, classnames etc, but currently it doesnt look like that. This is because you need to call the `asFragment` method in the expect statement:
```
    expect(asFragment()).toMatchSnapshot()
```
If you change this as re run the test, the existing snapshot will fail and it will prompt you to inspect the changes and update the snapshot or the component accordingly. In this case you want to update the snapshot by pressing `u`

File: src/__tests__/Search.test.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { fireEvent, getByRole, render, screen } from "@testing-library/react";
```
Comment (StuartCowley) [inline src/__tests__/Search.test.js:1]: You can remove most of these imports here as you are not currently using them in the tests:
```
import { render } from "@testing-library/react";
```

File: src/__tests__/App.test.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import { render, screen } from "@testing-library/react";
+import App from "../components/App";
+
+describe('App', () => {
+  const { asFragment } = render(<App />)
+
+  test("app renders correctly", () => {
```
Comment (StuartCowley) [inline src/__tests__/App.test.js:7]: Normally you would render the component inside each test block, could easily do this here:
```
  test("app renders correctly", () => {
      const { asFragment } = render(<App />)
```

File: src/__tests__/Search.test.js [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+import Search from "../components/Search";
+
+describe('Search Box', () => {
+  const { asFragment } = render(<Search />)
+
+  it("search box renders correctly", () => {
+    expect(asFragment).toMatchSnapshot()
+  })
```
Comment (StuartCowley) [inline src/__tests__/Search.test.js:9]: As above, should move this inside the `it` block:
```
  it("search box renders correctly", () => {
    const { asFragment } = render(<Search />)
    expect(asFragment).toMatchSnapshot()
  })
```

File: src/__tests__/Search.test.js [hunk 11]
```diff
@@ -0,0 +1,11 @@
+import { fireEvent, getByRole, render, screen } from "@testing-library/react";
+import Search from "../components/Search";
+
+describe('Search Box', () => {
+  const { asFragment } = render(<Search />)
```
Comment (StuartCowley) [inline src/__tests__/Search.test.js:5]: There is a warning in the console about required props being missing here. You should create a `validProps` object to house mocked versions of these to pass into the tests:
```
  const validProps = {
    setSearchResults: jest.fn(),
  };
```
Passing them in:
```
  it("search box renders correctly", () => {
    const { asFragment } = render(
      <Search setSearchResults={validProps.setSearchResults} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
```

Also, could some more tests be added here? You could assert that the function you are passing in is called when the search button is clicked for example

File: src/__tests__/SearchResults.test.js [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+describe('Search Results', () => {
+  const validProps = {
+    results: [],
+  }  
+  const { getByText } = render(<SearchResults results={validProps.results} />);
+
+  test("search results render correctly", () => {
+    expect(screen.getByText(/results/i)).toBeInTheDocument()
+  })
```
Comment (StuartCowley) [inline src/__tests__/SearchResults.test.js:13]: So a few things here: 
- You should put the render method inside the body the test as that is a thing that almost always needs to happen per test, to keep things separate for all the functionality you are trying to test
- You dont need to to desrtucture the `getByText` method from `render`, only to then get the same method off the screen object as you do on line 12. You could just render the component and remove the destructuring
- This isnt really much a useful test as it is. It passes, but only because you are looking for a value that will always be in the document regardless of there being any results to display or not. If you are passing in an empty array (as you are here), you would expect the message to be "No results" so you should assert for that:

```
  test("search results render correctly", () => {
    render(<SearchResults results={validProps.results} />);
    expect(screen.getByText("No results")).toBeInTheDocument()
  })
```
You now have half of the possible expected outcomes covered by your test. The other is what should be displayed if you pass an array in? Normally we would expect there to be the same number of images as the length of the array of URLs passed in, but your app is a bit more complex with an object being passed in and thumbnails/large images being displayed too, so this would need to be a bit more involved:

```
  test("Correct number of images are rendered with correct attributes", () => {
    validProps = {
      results: [
        {
          description: "test description 1",
          original: "https://test-original-url-1.jpg",
          originalHeight: 600,
          originalTitle: "test title 1",
          thumbnail: "https://test-thumbnail-url-1.jpg"
        },
        {
          description: "test description 2",
          original: "https://test-original-url-2.jpg",
          originalHeight: 800,
          originalTitle: "test title 2",
          thumbnail: "https://test-thumbnail-url-2.jpg"
        },
      ]
    };

    render(<SearchResults results={validProps.results} />)
    const allImages = screen.getAllByRole("img");

    expect(
      screen.getByText(`${validProps.results.length} results`)
    ).toBeInTheDocument();
    expect(allImages).toHaveLength(validProps.results.length * 2);
    expect(allImages[2]).toHaveAttribute("src", "https://test-original-url-1.jpg");
    expect(allImages[3]).toHaveAttribute("src", "https://test-original-url-2.jpg");
  });
```

Once you get the hang of it testing really can be quite powerful and an extremely useful tool to have at your disposal!

File: src/components/App.jsx [hunk 13]
```diff
@@ -0,0 +1,17 @@
+import { React, useState, useEffect } from "react";
```
Comment (StuartCowley) [inline src/components/App.jsx:1]: There is no useEffect being used here, you can remove this

File: src/components/SearchResults.jsx [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+
+SearchResults.defaultProps = {
+  length: 0,
+}
+
+SearchResults.propTypes = {
+  results: PropTypes.arrayOf (
+    PropTypes.object.isRequired,
+  ),
```
Comment (StuartCowley) [inline src/components/SearchResults.jsx:33]: This really should be a bit more detailed. This is especially important for your app is its a bit more complex than the track specifies and uses more complex data, this would be the perfect place to build out what that should look like

File: src/components/SearchResults.jsx [hunk 15]
```diff
@@ -0,0 +1,36 @@
+import { React, useState } from "react";
```
Comment (StuartCowley) [inline src/components/SearchResults.jsx:1]: You dont need to import state here as it isnt being used ion the component

File: src/requests/getImages.js [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+const getImages = (query) => {
+  if(!query) {
+    return Promise.resolve([])
+  } else {
+    return axios
+      .get(`https://images-api.nasa.gov/search?q=${query}`)
+      .then((response) => {
+        const imageResults = (response.data.collection.items)
+
+// ORIGINAL PARSE RESULTS METHOD
```
Comment (StuartCowley) [inline src/requests/getImages.js:12]: You dont need to include previous iterations to show you you got where you ended up, thats one of the main things version control is for. You can delete all this commented out code

File: src/requests/getImages.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+// ORIGINAL PARSE RESULTS METHOD         
+//
+//        const filteredResults = imageResults.filter(result => result.data[0].media_type === "image")
+//        console.log(imageResults)
+//        const images = filteredResults.map(result => result.links[0].href)
+
+//        return images
+
+        let results=[]
```
Comment (StuartCowley) [inline src/requests/getImages.js:20]: Could this be a const? Since you arent directly reassigning it theres no need to use the let keyword here

**Review by StuartCowley — COMMENTED** [review]
So this is pretty good! Youve got a really nice desktop app here, works as expected, and displays some real attention to detail. If left some info here for you about several areas, but in summary:

- What is the api.png file here for? Can it be removed?
- The readme is missing a few bits,
- A few suggestions for updating parts of the CSS,
- Updating proptypes to be more useful
- Theres quite a few issues in the tests, snapshots being out of date, functions not covered, required props missed off, that sort of thing. Have put some suggestions in there for you to fix some of the broken stuff and some things you could add to greatly increase the tests value as a whole

Thats pretty much it, probably not much more than an hours worth of work at most to update everything. Would be great to see the test coverage in particular improve as the app itself really looks good!

File: README.md (hunk 1 above)
Comment (billsongames) [reply README.md:1]: I did read a bit about README formatting but it's something I need to work on. The actual view looks ok so assumed the formatting was correct.

File: _code/code.js (hunk 4 above)
Comment (billsongames) [reply _code/code.js:1]: its old code that I kept for reference in case I needed it. Have added the folder to the .gitignore file

File: src/styles/searchResults.css (hunk 5 above)
Comment (billsongames) [reply src/styles/searchResults.css:1]: Fair point. classname used now instead

File: src/__tests__/SearchResults.test.js (hunk 12 above)
Comment (billsongames) [reply src/__tests__/SearchResults.test.js:13]: I think I was that involved in looking at the API and extracting the various bits of information from it and then inserting them into the ImageGallery component and testing it "manually" by checking it rendered correctly visually, that the writing of test code went out the window. That said, i do appreciate the importance of TDD and how it will be essential when doing a "real" tech test when it comes to applying for jobs. Have expanded the PropTypes for this test and test still passes :o)

File: src/components/SearchResults.jsx (hunk 15 above)
Comment (billsongames) [reply src/components/SearchResults.jsx:1]: probably left over from a previous version of the code where state was used.

File: src/requests/getImages.js (hunk 17 above)
Comment (billsongames) [reply src/requests/getImages.js:20]: I declared results using let as the array would be written to later in the code. Not necessary though as const is better to use
